### ISSUE TITLE
silent failure when accessing unbound variable with set -u

### ISSUE THREAD

**Issue by eedwards-sk** [bats-core/bats-core#81]
bash version: GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin17)

similar to #72 where toggling `set +/-e` during the command seems to be a workaround

test file

```
#!/usr/bin/env bats

setup() {
  set -u
}

@test "example" {
  echo $unbound_variable
}
```

output

```
→ ./test.bats
 ✗ example

1 test, 1 failure
```

with workaround

```
#!/usr/bin/env bats

setup() {
  set -u
}

@test "example" {
  set +e
  echo $unbound_variable
  set -e
}
```

workaround output

```
→ ./test.bats
 ✗ example
   (in test file test.bats, line 9)
     `echo $unbound_variable' failed
   /var/folders/dm/2_68m_h13h5dxq2f05wsc47c0000gq/T/bats.76966.src: line 9: unbound_variable: unbound variable

1 test, 1 failure
```

**Comment by tterranigma** [comment]
I cannot reproduce this with `GNU bash, version 3.2.51(1)-release (x86_64-unknown-linux-gnu)`. I get the stack trace as expected.

**Comment by mbland** [comment]
Rehashing my methodology from #72, here's the content of `unset.bash`:

```bash
#! /usr/bin/env bash

trap 'echo ERR:  status == "$?", funcname == "${FUNCNAME:-main}"' ERR
trap 'echo EXIT: status == "$?", funcname == "${FUNCNAME:-main}"' EXIT

do_it() {
  echo "$unset_var"
}
do_it
```

And as before, `/bin/bash` is 3.2.57(1)-release (stock Bash on macOS), and `/usr/local/bin/bash` is 4.4.19(1)-release:

```
$ /bin/bash -u unset.bash 
unset.bash: line 7: unset_var: unbound variable
EXIT: status == 1, funcname == do_it

$ /bin/bash -eu unset.bash 
unset.bash: line 7: unset_var: unbound variable
EXIT: status == 0, funcname == do_it

$ /usr/local/bin/bash -u unset.bash 
unset.bash: line 7: unset_var: unbound variable
EXIT: status == 1, funcname == main

$ /usr/local/bin/bash -eu unset.bash 
unset.bash: line 7: unset_var: unbound variable
EXIT: status == 1, funcname == do_it
```

Interesting things to note:

* The `ERR` trap is _never_ invoked. This is because the error is during parameter expansion, before any expression evaluation is attempted.
* Updating the expression to `echo "$unset_var" || return 1` didn't change the output _at all_. Again, this is because the error happened during parameter expansion, before even evaluating the expression.
* The only time `do_it` returns is under 4.4.19(1)-release _without_ `set -e`.
* Under 3.2.57(1)-release, the exit status under `set -e` is zero, just like it was with `source "nonexistent file"` from #72.

Chapter and verse from https://tiswww.case.edu/php/chet/bash/CHANGES:

```
This document details the changes between this version, bash-4.1-alpha,
and the previous version, bash-4.0-release.

1.  Changes to Bash

x.  Fixed a bug that caused $? to not be set correctly when referencing an
    unset variable with set -u and set -e enabled.
```

Now I'm wondering if that same fix might've affected the exit status of `source "nonexistent file"` as well. Will definitely have to take a look at the source and pinpoint the change when I get a chance. (I've got a system set up on my home machine to do this, but only have my work machine at the moment.)

With regard to Bats, note that `run()` does run `set +e` before executing the command to be evaluated. You should be able to set `set -u` and have `run()` capture unset variable errors in `$output`. (We made sure bats-core was itself `set -u`/strict-mode safe in #26.) In fact, we should

- [ ] add a test case to ensure `run()` captures unset variable errors in `$output` when `set -u` is enabled.

**Comment by eedwards-sk** [comment]
@tterranigma Thanks for trying to repo! I'm currently on Mac OS X 10.13.4 (High Sierra) with iTerm2 in case somehow those specifics are related to this failure case.

@tterranigma I was seeing the same issue when using `run` (in fact that's what first tipped me off, I thought it was `run` that was doing it).

```
#!/usr/bin/env bats

setup() {
  set -u
}

@test "example" {
  run echo $unbound_variable
}
```

```
→ ./test.bats
 ✗ example

1 test, 1 failure
```

**Comment by mbland** [comment]
Oh, I meant `set -u` shouldn't interfere when calling "run some-script-or-function". It doesn't surprise me that it failed when calling `run echo $unbound_variable`, because the error happens before `run` even gets a chance to execute.

Another reason I need to write a test case: Now that I think about it, I'm not sure that `set -u` will apply within the subshell launched by `run()`. We'll see. But at the immediate moment, my boss's boss just instigated a biergarten run, so I'd best comply. 😛

**Comment by mbland** [comment]
The following patch accounts for the changes between Bash 4.0.44 and Bash 4.1.0 that explain the `set -eu` behavior. I was able to generate it thanks to the following chapter and verse from https://tiswww.case.edu/php/chet/bash/CHANGES:

```
This document details the changes between this version, bash-4.1-alpha,
and the previous version, bash-4.0-release.

1.  Changes to Bash

x.  Fixed a bug that caused $? to not be set correctly when referencing an
    unset variable with set -u and set -e enabled.
```

Which corresponds to the following lines from bash-4.1.0/ChangeLog:

```
				 1/2/2009
				 --------

[...snip...]

				   2/20
				   ----

[bash-4.0 released]

[...snip...]

				   4/10
				   ----
{expr,subst}.c
	- make sure last_command_exit_value is set to EXECUTION_FAILURE
	  before calling err_unboundvar, in case set -e is enabled and
	  the shell exits from there.  Fixes bug reported by Freddy
	  Vulto <fvulto@gmail.com> and Piotr Zielinski
	  <piotr.zielinski@gmail.com>

[bash-4.1-alpha frozen]
```

What it does is ensure that the `last_command_exit_value` variable is set to `EXECUTION_FAILURE` (i.e. 1) _before_ calling `err_unboundvar()`, which will exit immediately when `set -e` is in effect. This ensures that `$?` is nonzero such that when `bats_teardown_trap` calls `bats_error_trap`, it properly captures the stack trace pointing to the failing line, even though the `ERR` trap isn't triggered by Bash itself. (Note that I snipped the diff for `subst.c` to focus only on the changes pertaining `last_command_exit_value`.)

```diff
--- bash-4.0.44/expr.c	2009-01-04 14:32:29.000000000 -0500
+++ bash-4.1.0/expr.c	2009-08-05 19:19:25.000000000 -0400
@@ -202,7 +202,7 @@
 static int expr_stack_size;	   /* Number of slots already allocated. */
 
 extern char *this_command_name;
-extern int unbound_vars_is_error;
+extern int unbound_vars_is_error, last_command_exit_value;
 
 #if defined (ARRAY_VARS)
 extern const char * const bash_badsub_errmsg;
@@ -923,6 +923,7 @@
       value = tok;
 #endif
 
+      last_command_exit_value = EXECUTION_FAILURE;
       err_unboundvar (value);
 
 #if defined (ARRAY_VARS)
--- bash-4.0.44/subst.c	2017-03-19 12:22:07.000000000 -0400
+++ bash-4.1.0/subst.c	2009-12-30 08:24:28.000000000 -0500
@@ -5058,6 +5253,7 @@
     {
       c = *--t;
       *t = '\0';
+      last_command_exit_value = EXECUTION_FAILURE;
       err_unboundvar (s);
       *t = c;
       return (-1);
@@ -6929,8 +7131,8 @@
 	  uerror[0] = '$';
 	  uerror[1] = c;
 	  uerror[2] = '\0';
-	  err_unboundvar (uerror);
 	  last_command_exit_value = EXECUTION_FAILURE;
+	  err_unboundvar (uerror);
 	  return (interactive_shell ? &expand_wdesc_error : &expand_wdesc_fatal);
 	}
       if (temp1)
@@ -6977,8 +7179,8 @@
 	      uerror[0] = '$';
 	      uerror[1] = c;
 	      uerror[2] = '\0';
-	      err_unboundvar (uerror);
 	      last_command_exit_value = EXECUTION_FAILURE;
+	      err_unboundvar (uerror);
 	      return (interactive_shell ? &expand_wdesc_error : &expand_wdesc_fatal);
 	    }
 	}
@@ -7280,7 +7491,10 @@
 
 unbound_variable:
       if (unbound_vars_is_error)
-	err_unboundvar (temp1);
+	{
+	  last_command_exit_value = EXECUTION_FAILURE;
+	  err_unboundvar (temp1);
+	}
       else
 	{
 	  free (temp1);
```

I did build a version of Bash 4.0.44 with the above patch applied to verify this analysis. Again, using the test script:

```bash
#! /usr/bin/env bash

trap 'echo ERR:  status == "$?", funcname == "${FUNCNAME:-main}"' ERR
trap 'echo EXIT: status == "$?", funcname == "${FUNCNAME:-main}"' EXIT

do_it() {
  echo "$unset_var"
}
do_it
```

The two builds produce the following results:

```
$ 4.0.44/bin/bash -u unbound.test
unbound.test: line 7: unset_var: unbound variable
EXIT: status == 1, funcname == do_it

$ 4.0.44/bin/bash -eu unbound.test
unbound.test: line 7: unset_var: unbound variable
EXIT: status == 0, funcname == do_it

$ 4.0.44-with-unbound-fix/bin/bash -u unbound.test
unbound.test: line 7: unset_var: unbound variable
EXIT: status == 1, funcname == do_it

$ 4.0.44-with-unbound-fix/bin/bash -eu unbound.test
unbound.test: line 7: unset_var: unbound variable
EXIT: status == 1, funcname == do_it
```

I'd speculated that this patch might've resolved the issue from #72 regarding the inconsistent behavior of `source "nonexistent file"`, but the two versions produce the same results in that case. I'll expound more in that issue.

I'll have to investigate and report separately regarding why the unbound variable would exit the shell immediately in older versions of Bash even when `set -e` wasn't in effect (as opposed to causing the enclosing function to return an error, as it appears to in Bash 4.4.19(1)-release based on the previously posted results).

Also, I still need to try adding a test case to see how `set -u` affects `run()`, but at least we now know exactly what's causing the issue—and that there's nothing Bats itself can really do when a test file triggers the issue in Bash versions before 4.1.0.

**Comment by mbland** [comment]
@eedwards-sk All is not lost; #82 resolves the `set -u` problem in `setup()` and in test case functions. (It doesn't resolve it for `teardown()`, but that may come in a future PR.) Only arrived at that solution after all the investigation and thinking on things a bit.

**Comment by mbland** [comment]
Closed by #82.
